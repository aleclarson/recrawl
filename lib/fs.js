Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
exports.localFs = {
    readdir: async (name) => fs.readdirSync(name),
    readlink: async (name) => fs.readlinkSync(name),
    lstat: async (name) => fs.lstatSync(name),
    stat: async (name) => fs.statSync(name),
};
