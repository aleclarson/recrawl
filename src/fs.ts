import * as fs from 'fs'

export interface FileAdapter {
  readdir(name: string): Promise<string[]>
  readlink(name: string): Promise<string>
  lstat(name: string): Promise<fs.Stats>
  stat(name: string): Promise<fs.Stats>
}

export const localFs: FileAdapter = {
  readdir: async name => fs.readdirSync(name),
  readlink: async name => fs.readlinkSync(name),
  lstat: async name => fs.lstatSync(name),
  stat: async name => fs.statSync(name),
}
