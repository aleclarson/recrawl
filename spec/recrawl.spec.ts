import * as fs from 'fs'
import * as path from 'path'
import { crawl } from '..'

const root = path.join(__dirname, '__fixtures__', 'root')

it('collects all descendants when no arguments are passed', () => {
  expect(crawl(root)).toMatchSnapshot()
})

it('treats directory symlinks like real directories', () => {
  const z = path.join(root, 'z')
  expect(fs.readlinkSync(z)).toBe('a')

  const entered = []
  crawl(root, { enter: dir => (entered.push(dir), true) })
  expect(entered).toContain('z')
})

const isWindows =
  process &&
  (process.platform === 'win32' || /^(msys|cygwin)$/.test(process.env.OSTYPE))

isWindows &&
  it('always returns unix paths', () => {
    crawl(root).forEach(file => {
      expect(file.indexOf('\\')).toBe(-1)
    })
  })

describe('options.only', () => {
  it('has no effect when empty', () => {
    expect(crawl(root, { only: [] })).toEqual(crawl(root))
  })
  it('is applied to files', () => {
    const paths = crawl(root, { only: ['*.js'] })
    expect(paths).toMatchSnapshot()
    expect(paths).not.toContain('node_modules/foo/package.json')
  })
  it('is applied to symlinks', () => {
    expect(fs.readlinkSync(path.join(root, 'index.js'))).toBe('a/b/c/3.js')
    const paths = crawl(root, { only: ['index.js'] })
    expect(paths).toEqual(['index.js'])
  })
  it('is not applied to directories', () => {
    const paths = crawl(root, { only: ['a', 'z'] })
    expect(paths).toEqual([])
  })
})

describe('options.skip', () => {
  it('has no effect when empty', () => {
    expect(crawl(root, { skip: [] })).toEqual(crawl(root))
  })
  it('is applied to files', () => {
    const paths = crawl(root, { skip: ['*.js'] })
    expect(paths).toMatchSnapshot()
    expect(paths).not.toContain('a/1.js')
  })
  it('is applied to symlinks', () => {
    const paths = crawl(root, { skip: ['*.js'] })
    expect(paths).not.toContain('index.js')
  })
  it('is applied to directories', () => {
    const paths = crawl(root, { skip: ['node_modules'] })
    expect(paths).not.toContain('node_modules/foo/package.json')
  })
  it('overrides the "only" option', () => {
    const paths = crawl(root, { only: ['a'], skip: ['a'] })
    expect(paths).toEqual([])
  })
})

describe('options.deep', () => {
  it('skips directories when false', () => {
    const paths = crawl(root, { deep: false })
    expect(paths).toMatchSnapshot()
  })
})

describe('options.depth', () => {
  it('limits the traversal depth', () => {
    const paths = crawl(root, { depth: 1 })
    expect(paths).toMatchSnapshot()
  })
  it('skips directories when zero', () => {
    const paths = crawl(root, { depth: 0 })
    expect(paths).toMatchSnapshot()
  })
})

describe('options.enter', () => {
  it('is called before entering a directory', () => {
    const enter = jest.fn(() => true)
    crawl(root, { enter })
    expect(enter.mock.calls).toMatchSnapshot()
  })
  it('can return false to skip a directory', () => {
    const enter = jest.fn(() => false)
    const paths = crawl(root, { enter })
    expect(enter.mock.calls).toMatchSnapshot()
    expect(paths).toMatchSnapshot()
  })
})

describe('options.filter', () => {
  it('is called whenever a file or symlink is found', () => {
    const filter = jest.fn(() => true)
    crawl(root, { filter })
    expect(filter.mock.calls).toMatchSnapshot()
  })
  it('is called after the "only" option is applied', () => {
    const filter = jest.fn(() => true)
    crawl(root, { filter, only: ['*.js'] })
    expect(filter.mock.calls).toMatchSnapshot()
  })
  it('is called after the "skip" option is applied', () => {
    const filter = jest.fn(() => true)
    crawl(root, { filter, skip: ['*.js'] })
    expect(filter.mock.calls).toMatchSnapshot()
  })
})

describe('options.follow', () => {
  it('resolves every symlink when true', () => {
    const paths = crawl(root, { follow: true })
    expect(paths['index.js']).toBe('a/b/c/3.js')
  })
  it('is called whenever a symlink is found', () => {
    const follow = jest.fn(() => true)
    crawl(root, { follow })
    expect(follow.mock.calls).toMatchSnapshot()
  })
  it('can return false to use the symlink path instead of resolving it', () => {
    const paths = crawl(root, { follow: () => false })
    expect(paths['index.js']).toBe(true)
  })
  it('avoids resolving symlinks when zero', () => {
    const paths = crawl(root, { follow: 0 })
    expect(paths['index.js']).toBe(true)
  })
  it.todo('limits the link depth when > zero')
  it.todo('can resolve symlinks to symlinks')
  it.todo('detects symlink infinite recursion')
})

describe('options.adapter', () => {
  it.todo('lets you provide your own filesystem')
})
