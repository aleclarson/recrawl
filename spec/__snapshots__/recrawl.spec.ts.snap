// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collects all descendants when no arguments are passed 1`] = `
Array [
  ".vscode/settings.json",
  "a/1.js",
  "a/b/2.js",
  "a/b/c/3.js",
  "index.js",
  "node_modules/foo/package.json",
  "z/1.js",
  "z/b/2.js",
  "z/b/c/3.js",
]
`;

exports[`options.deep skips directories when false 1`] = `
Array [
  "index.js",
]
`;

exports[`options.depth limits the traversal depth 1`] = `
Array [
  ".vscode/settings.json",
  "a/1.js",
  "index.js",
  "z/1.js",
]
`;

exports[`options.depth skips directories when zero 1`] = `
Array [
  "index.js",
]
`;

exports[`options.enter can return false to skip a directory 1`] = `
Array [
  Array [
    ".vscode",
    0,
  ],
  Array [
    "a",
    0,
  ],
  Array [
    "node_modules",
    0,
  ],
  Array [
    "z",
    0,
  ],
]
`;

exports[`options.enter can return false to skip a directory 2`] = `
Array [
  "index.js",
]
`;

exports[`options.enter is called before entering a directory 1`] = `
Array [
  Array [
    ".vscode",
    0,
  ],
  Array [
    "a",
    0,
  ],
  Array [
    "a/b",
    1,
  ],
  Array [
    "a/b/c",
    2,
  ],
  Array [
    "node_modules",
    0,
  ],
  Array [
    "node_modules/foo",
    1,
  ],
  Array [
    "z",
    0,
  ],
  Array [
    "z/b",
    1,
  ],
  Array [
    "z/b/c",
    2,
  ],
]
`;

exports[`options.filter is called after the "only" option is applied 1`] = `
Array [
  Array [
    "a/1.js",
    "1.js",
  ],
  Array [
    "a/b/2.js",
    "2.js",
  ],
  Array [
    "a/b/c/3.js",
    "3.js",
  ],
  Array [
    "index.js",
    "index.js",
  ],
  Array [
    "z/1.js",
    "1.js",
  ],
  Array [
    "z/b/2.js",
    "2.js",
  ],
  Array [
    "z/b/c/3.js",
    "3.js",
  ],
]
`;

exports[`options.filter is called after the "skip" option is applied 1`] = `
Array [
  Array [
    ".vscode/settings.json",
    "settings.json",
  ],
  Array [
    "node_modules/foo/package.json",
    "package.json",
  ],
]
`;

exports[`options.filter is called whenever a file or symlink is found 1`] = `
Array [
  Array [
    ".vscode/settings.json",
    "settings.json",
  ],
  Array [
    "a/1.js",
    "1.js",
  ],
  Array [
    "a/b/2.js",
    "2.js",
  ],
  Array [
    "a/b/c/3.js",
    "3.js",
  ],
  Array [
    "index.js",
    "index.js",
  ],
  Array [
    "node_modules/foo/package.json",
    "package.json",
  ],
  Array [
    "z/1.js",
    "1.js",
  ],
  Array [
    "z/b/2.js",
    "2.js",
  ],
  Array [
    "z/b/c/3.js",
    "3.js",
  ],
]
`;

exports[`options.follow is called whenever a symlink is found 1`] = `
Array [
  Array [
    "index.js",
    0,
  ],
]
`;

exports[`options.follow resolves every symlink when true 1`] = `
Object {
  ".vscode/settings.json": true,
  "a/1.js": true,
  "a/b/2.js": true,
  "a/b/c/3.js": true,
  "index.js": "b/c/3.js",
  "node_modules/foo/package.json": true,
  "z/1.js": true,
  "z/b/2.js": true,
  "z/b/c/3.js": true,
}
`;

exports[`options.only is applied to files 1`] = `
Array [
  "a/1.js",
  "a/b/2.js",
  "a/b/c/3.js",
  "index.js",
  "z/1.js",
  "z/b/2.js",
  "z/b/c/3.js",
]
`;

exports[`options.skip is applied to files 1`] = `
Array [
  ".vscode/settings.json",
  "node_modules/foo/package.json",
]
`;
